src/bootstrap/download.rs

```diff
@@ -367,26 +367,74 @@ impl Config {
 
     pub(crate) fn download_ci_rustc(&self, commit: &str) {
         self.verbose(&format!("using downloaded stage2 artifacts from CI (commit {commit})"));
+
         let version = self.artifact_version_part(commit);
+        // download-rustc doesn't need its own cargo, it can just use beta's. But it does need the
+        // `rustc_private` crates for tools.
+        let extra_components = ["rustc-dev"];
+
+        self.download_toolchain(
+            &version,
+            "ci-rustc",
+            commit,
+            &extra_components,
+            Self::download_ci_component,
+        );
+    }
+
+    pub(crate) fn download_beta_toolchain(&self) {
+        self.verbose(&format!("downloading stage0 beta artifacts"));
+
+        let date = &self.stage0_metadata.compiler.date;
+        let version = &self.stage0_metadata.compiler.version;
+        let extra_components = ["cargo"];
+
+        let download_beta_component = |config: &Config, filename, prefix: &_, date: &_| {
+            config.download_component(DownloadSource::Dist, filename, prefix, date, "stage0")
+        };
+
+        self.download_toolchain(
+            version,
+            "stage0",
+            date,
+            &extra_components,
+            download_beta_component,
+        );
+    }
+
+    fn download_toolchain(
+        &self,
+        // FIXME(ozkanonur) use CompilerMetadata instead of `version: &str`
+        version: &str,
+        sysroot: &str,
+        stamp_key: &str,
+        extra_components: &[&str],
+        download_component: fn(&Config, String, &str, &str),
+    ) {
+        if env::var_os("BOOTSTRAP_PYTHON").is_some() {
+            return
+        }
```

Comment:
This PR forms a part of the `remove/replace python scripts from bootstrap`. Its objective is to download stage0 directly from the Rust source, rather than through the Python script. However, as we are still utilizing Python scripts currently, stage0 is downloaded twice. The second attempt involves executing the downloaded stage0 compiler, which is incompatible with Windows FS and causes an error, as can be seen in [this error](https://github.com/rust-lang-ci/rust/actions/runs/4591544753/jobs/8107789452#step:26:30881).


This check allows us to determine if bootstrap python involved or not(which also means stage0 is already downloaded). So we can prevent the removal of directories/files that are currently in use.

@jyn514 Please correct me if I am wrong since you are the author of this implementation.

Reply:
I don't think this check is correct. Imagine that we implement https://github.com/rust-lang/rust/issues/99989 in the python script. Then python will download only bootstrap-shim and not stage0, but `BOOTSTRAP_PYTHON` will still be set, so we'll get an error when we try to use stage0 to build std.

I would *expect* us to automatically avoid downloading the compiler if it already exists on disk, because of the `rustc_stamp` check. I think there is some different between the python stamp and the rust stamp on disk. Can you check and see if it's overwriting an existing stamp, maybe by adding `println!("{}", std::fs::read_to_string(rustc_stamp))` on line 420?

Reply:
Sometimes we don't even create stamp file while bootstrap tests are performed.

I just discovered comparing stamp doesn't do anything wrong, checking the `rustc` executable what makes the tests fail. Is there any specific case that requires `!bin_root.join("bin").join("rustc").exists()` check? Tested on windows, when I remove this check from the if condition, all goes well.

Reply:
LOL, it's because we are missing `.exe` extension on windows :cold_face: